### ISSUE TITLE
Consider introducing a `LanguageIdentifier` trait

### ISSUE THREAD

**Issue by zbraniecki** [zbraniecki/unic-locale#13]
Currently, we have a `LanguageIdentifier` struct and a strong conversion from `Locale` for it.

That allows customers to say "I want to accept anything that works like LanguageIdentifier" which is useful to allow code to use `langid` and later switch to `locale` if needed or even use some custom structs that act like one.

A good example of where such code can be useful is language negotiation, where we want to negotiate between lists of anything that can be treated as LanguageIdentifier.

In the conversation there, we started talking that maybe what we really need is a trait that anything can implement and any code that needs language management should just accept things that implement that trait.

So, for example:

```rust
fn negotiate_languages<
  R: LangIdTrait,
  A: LangIdTrait
>(requested: &[R], available: &[A]) -> Vec<&A>;

struct PluralRules {
  fn new(lang: LangIdTrait);
}

struct FluentBundle {
  fn new(lang: LangIdTrait);
}
```

If that's the case, we'd need to pick a name (`LanguageIdentifier` for trait? struct? what about the other?), and rework this code a bit to make the struct (auto)implement the trait.

**Comment by zbraniecki** [comment]
@Manishearth @cmyr @raphlinus - moving the conv. here.

I have no experience with designing traits vs. structs for general purpose such as carrying around language identifier info. Can you advise whether you think it's a good idea?
What would we do if we wanted to have CLDR Language Identifier API needs for the standard library?

**Comment by zbraniecki** [comment]
As an experiment I created a branch `trait` here and in fluent-locale and implemented the traits. I'm a bit concerned that we're going to end up copying/cloning much more, but so far the benchmarks don't indicate any perf regression.

I'd appreciate someone taking a look at those two PRs and verifying that it looks reasonable.

One change I made is that I now consume the structs in negotiate_languages, rather than references to them. I'm not sure if it's the right move - I believe we don't need them, just references to them, but I failed to figure out how to make the API signature work for a list of borrows of structs that implement a trait.
